src/tactic/push_neg.lean

```diff
@@ -0,0 +1,146 @@
+theorem not_exists_eq : (¬ ∃ x, s x) = (∀ x, ¬ s x) := propext not_exists
+theorem not_implies_eq : (¬ (p → q)) = (p ∧ ¬ q) := propext not_imp
+
+theorem classical.implies_iff_not_or : (p → q) ↔ (¬ p ∨ q) := imp_iff_not_or
+
+
+theorem not_eq (a b : α) : (¬ a = b) ↔ (a ≠ b) := iff.rfl
+
+variable  {β : Type u}
+variable [linear_order β]
+theorem not_le_eq (a b : β) : (¬ (a ≤ b)) = (b < a) := propext not_le
+theorem not_lt_eq (a b : β) : (¬ (a < b)) = (b ≤ a) := propext not_lt
+end
+
+meta def whnf_reducible (e : expr) : tactic expr := whnf e reducible
+
+private meta def transform_negation_step (e : expr) :
+  tactic (option (expr × expr)) :=
+do e ← whnf_reducible e,
+   match e with
+   | `(¬ %%ne) :=
+      (do ne ← whnf_reducible ne,
+      match ne with
+      | `(¬ %%a)      := do pr ← mk_app ``not_not_eq [a],
+                            return (some (a, pr))
+      | `(%%a ∧ %%b)  := do pr ← mk_app ``not_and_eq [a, b],
+                            return (some (`(¬ %%a ∨ ¬ %%b), pr))
+      | `(%%a ∨ %%b)  := do pr ← mk_app ``not_or_eq [a, b],
+                            return (some (`(¬ %%a ∧ ¬ %%b), pr))
+      | `(%%a ≤ %%b)  := do e ← to_expr ``(%%b < %%a),
+                            pr ← mk_app ``not_le_eq [a, b],
+                            return (some (e, pr))
+      | `(%%a < %%b)  := do e ← to_expr ``(%%b ≤ %%a),
+                            pr ← mk_app ``not_lt_eq [a, b],
+                            return (some (e, pr))
+      | `(Exists %%p) := do pr ← mk_app ``not_exists_eq [p],
+                            e ← match p with
+                                | (lam n bi typ bo) := do
+                                    body ← mk_app ``not [bo],
+                                    return (pi n bi typ body)
+                                | _ := tactic.fail "Unexpected failure negating ∃"
+                                end,
+                            return (some (e, pr))
+      | (pi n bi d p) := if p.has_var then do
+                            pr ← mk_app ``not_forall_eq [lam n bi d p],
+                            body ← mk_app ``not [p],
+                            e ←  mk_app ``Exists [lam n bi d body],
+                            return (some (e, pr))
+                         else do
+                            pr ← mk_app ``not_implies_eq [d, p],
+                            `(%%_ = %%e') ← infer_type pr,
+                            return (some (e', pr))
+      | _             := return none
+      end)
+    | _        := return none
+  end
+
+private meta def transform_negation : expr → tactic (option (expr × expr)) :=
+λ e, do
+  opr ← transform_negation_step e,
+  match opr with
+  | (some (e', pr)) := do
+    opr' ← transform_negation e',
+    match opr' with
+    | none              := return (some (e', pr))
+    | (some (e'', pr')) := do pr'' ← mk_eq_trans pr pr',
+                              return (some (e'', pr''))
+    end
+  | none            := return none
+  end
```

Comment:
I think you could rewrite this more straightforwardly as:

```lean
private meta def transform_negation : expr → tactic (option (expr × expr))
| e :=
do (some (e', pr)) ← transform_negation_step e | return none,
   (some (e'', pr')) ← transform_negation e' | return (some (e', pr)),
   pr'' ← mk_eq_trans pr pr',
   return (some (e'', pr''))
```

Reply:
Thanks. This code was copied from `finish` (I only removed a configuration variable). I'll try your version.

Reply:
Done